### ISSUE TITLE
Fuzzer tests need instructions on local reproduction

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#7552]
The tests under fuzz/ only seem to be usable in conjunction with a nontrivial harness, such as OSS-Fuzz, or something similar; when bugs are inevitably reported, we need a way to reproduce them locally to diagnose and fix them, but it's not clear how to do this. We need some documentation (at least) for how to address this.

(Note that I am able to diagnose some that are found internally at Google in our fuzz-testing framework, but this will not be useful to the many Halide contributors who don't work at Google.)

attn @silvergasp

**Comment by silvergasp** [comment]
> The tests under fuzz/ only seem to be usable in conjunction with a nontrivial harness, such as OSS-Fuzz, or something similar; when bugs are inevitably reported, we need a way to reproduce them locally to diagnose and fix them, but it's not clear how to do this. We need some documentation (at least) for how to address this.

I agree that there should be some documentation here, and I'm happy to help out with. However I actually think that in terms of reproducing issues, it's going to be a lot easier to just use the existing infrastructure around OSS-fuzz.

> (Note that I am able to diagnose some that are found internally at Google in our fuzz-testing framework, but this will not be useful to the many Halide contributors who don't work at Google.)

There is [extensive documentation on how to use some of the oss-fuzz infrastructure locally](https://google.github.io/oss-fuzz/getting-started/new-project-guide/#testing-locally), that doesn't require you to be a Google employee (I don't personally work at Google). But I agree that the Halide project could benefit from some project specific documentation, even if it's as simple as something like.
```
git clone https://github.com/google/oss-fuzz.git
cd oss-fuzz
export PROJECT_NAME=halide
python infra/helper.py build_image $PROJECT_NAME
python infra/helper.py build_fuzzers --sanitizer address --engine libfuzzer $PROJECT_NAME
# Reproduce a test case
python infra/helper.py reproduce $PROJECT_NAME <fuzz_target_name> <testcase_path>
```
I think that perhaps documenting this process would be far more beneficial as many of the bugs (if there are any :P) that will be found will be found by X fuzzing engine with Y sanitizer. It's much easier to automate reproduction with all of the open source OSS-fuzz infrastructure than it is to create your own AFL->Libfuzzer adapters etc. Not to mention that there are going to be some subtle differences between a regular build that in some cases *may* make reproduction more challenging (e.g. linking against libc++ instead of stdlibC++, or ASAN sanitizer instrumentation of the entire llvm library).

What are your thoughts on just documenting the project specific aspects of oss-fuzz integration/local testing? I'm happy to open a PR with something along those lines if that sounds reasonable too you?

**Comment by abadams** [comment]
I'm pretty concerned with the situation as it stands. I may be wrong in some details, but a pessimistic view of the recent fuzzing work is that it took a bunch of perfectly good tests, removed them from CI, and added a heavy dependency if you want to run them at all, but there are no instructions in the repo for how to do it. 

We need to get these tests running again on CI, and we need an easy way for Halide devs with no experience with oss-fuzz to reproduce and fix failures when reported on CI or elsewhere.

**Comment by steven-johnson** [comment]
Yeah, I agree with Andrew -- I should have reviewed the original PRs more carefully to realize the implications.

From the documentation about how to use oss-fuzz locally: `You can build your docker image and fuzz targets locally ... You *must* run your fuzz target binaries inside the base-runner docker container to make sure that they work properly ` -- if this requires using Docker locally to work, I have concerns; we don't currently have any Docker usage or requirements in our testing infrastructure, and adding one would be a heavyweight addition. 

It seems like the fuzzer API is pretty shallow -- surely there's some simple shim that exists that could just provide a stream of random numbers? (If not, I think we need to rethink this.)

**Comment by abadams** [comment]
So is the current situation that we can build the tests without an oss-fuzz environment, but we can't usefully run them without it?

**Comment by steven-johnson** [comment]
> So is the current situation that we can build the tests without an oss-fuzz environment, but we can't usefully run them without it?

Not as-is, because the need to include `#include <fuzzer/FuzzedDataProvider.h>`

**Comment by silvergasp** [comment]
> Not as-is, because the need to include #include <fuzzer/FuzzedDataProvider.h>

Is `fuzzer/FuzzedDataProvider.h` not available on mac? it was my impression that this was just shipped with clang. It is a [single-header library/utility](https://github.com/llvm/llvm-project/blob/main/compiler-rt/include/fuzzer/FuzzedDataProvider.h), so it could easily be vendored in/modified to meet Halides needs (e.g. add pick_value_in_vector).

**Comment by steven-johnson** [comment]
On Tue, May 9, 2023 at 10:53 AM Nathaniel Brough ***@***.***>
wrote:

> Not as-is, because the need to include #include
> <fuzzer/FuzzedDataProvider.h>
>
> Is fuzzer/FuzzedDataProvider.h not available on mac? it was my impression
> that this was just shipped with clang.
>
Clang does not ship on Windows, though.

**Comment by silvergasp** [comment]
> Clang does not ship on Windows, though.

I might be misunderstanding something here, but I'm fairly certain clang is available on windows... https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.2 . If you open the assets drop down and then search for `.exe` or `win64` its there. I remember using it a year or two ago for a project, but I've since uninstalled it. However I haven't checked to see if libfuzzer is shipped with the windows version.

**Comment by silvergasp** [comment]
I'll have to clean out some room on my hard-drive to reinstall all the windows SDK's so it might take me a while. But I'm happy to give it a go on windows.

**Comment by silvergasp** [comment]
> It seems like the fuzzer API is pretty shallow -- surely there's some simple shim that exists that could just provide a stream of random numbers? (If not, I think we need to rethink this.)

Could you elaborate this. I think I that you are correct, but I feel like there are at least a few ways I could interpret this question.

**Comment by steven-johnson** [comment]
No, sorry that I wasn't clear -- I am aware that Clang-for-Windows exists, but Halide doesn't use it (we use MSVC because that's what ~all of the downstream Windows users of Halide use). The Halide team has very limited testing and personnel at the moment and we definitely do not have the resources to even consider supporting building/testing Halide using Clang-for-Windows, regardless of how easy it might be, so any solution that requires that is a nonstarter.

Now, that said, it may well be that for fuzz testing, we would be fine skipping support on Windows -- that's already the case for several other things that are technically supported on Windows but that we skip in the name of conserving testing resources (e.g. WebAssembly, WebGPU). 

I think the real high-order bits of all the concerns here are:
- We need a way to continue to build and execute these fuzz-tests locally on (at least) Linux and macOS systems. This shouldn't involve any special tooling or commands -- running `CTest` should execute the tests locally, somehow.
- Requiring the use of Docker is not ideal, for reasons that aren't entirely important here.
- Being able to run the tests locally shouldn't mean that we can't run them under OSS-Fuzz or other hosted setup, because being able to use those is a huge testing advantage that we really are glad to have (and glad to have the contribution of, despite the negative tone of my comments :-) )
- When a fuzz error is reported from OSS-Fuzz, we need some way to replicate that run locally -- presumably by feeding the same random-data seed or whatnot as the input. Based on the docs you pointed at, it appears that this may require running the 'official' fuzzer harness, set up in Docker, etc., which, as mentioned above, is problematic.

> > It seems like the fuzzer API is pretty shallow
> Could you elaborate this.

Sure, what I meant was that (naively) it seems like writing a simple `main()` wrapper, that provides an implementation of `FuzzedDataProvider` that is tied to (say) `std::mt19937` to generate random data, would be a plausible way to run this locally. 

Unfortunately, while this approach is likely tractable for maintaining the same level of testing that we had in our fuzz tests prior to this change, it would be useless for trying to repeat failures reported by OSS-Fuzz (unless we could somehow replicate the exact RNG sequence being used upstream).

**Comment by silvergasp** [comment]
As a bit of an example of how the workflow might work for you with OSS-fuzz I'll use a [bug that was found in simdutf](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=57673) (now publicly disclosed).
1. Build the fuzzer locally
2. Download failing test case from oss-fuzz bug tracker
![image](https://github.com/halide/Halide/assets/7277663/7ef5c05d-65be-4849-b29e-060c521cc97e)
3. Run the fuzzer with the file you just downloaded.

`my_fuzzer downloaded_testcase`

I'm of course happy to document that process in markdown if that would help?

**Comment by silvergasp** [comment]
So I've been looking into all of this a bit more through the lens of trying to get this easier to use/develop with, and I've come up with a bit of a plan, that I'd like to check if it sane (I'm far from a Cmake wizard, so feel free to shoot me down if you don't think this is going to work).

- [ ] Create a cmake toolchain specifically for fuzzing similar to https://github.com/halide/Halide/blob/main/cmake/toolchain.linux-x64-asan.cmake
- [ ] Create a CMake preset for fuzzing.
- [ ] Create some docs on how to build all of these elements.

The only thing I'm unsure about here would be if you can combine to toolchains/presets so that you could do something like ASAN+Fuzzer or just Fuzzer alone?

**Comment by steven-johnson** [comment]
> The only catch with running the new libfuzzer harnesses with CTest would be that (by default) they will run for;
> 
> * An infinite amount of time OR
> * Until it finds a bug (ASAN errors/asserts) OR
> * You manually exit out of the test (ctrl-C).
> 
> Having said that, there are at least two ways to constrain the running time of the fuzz test e.g. using the flags `-max_total_time=<n seconds>` or `-runs=".

I presume that the OSS-Fuzz testing harness sets a limit of time or iterations, since it obviously doesn't run each test forever, so we can just mimic what they do (e.g. --runs=10).
 
> So the way you would normally reproduce a bug found with libfuzzer is to just pass file that contains the fuzzed data. e.g.

Yeah, that should work -- but I can't test it yet because I don't have a local configuration that will pass the HAS_FUZZ_FLAGS test.
 
> Create a cmake toolchain specifically for fuzzing

That sounds like a fantastic idea.

> if you can combine to toolchains/presets

I don't think so, but I'm not 100% certain; @alexreinking would know. (If not, it probably wouldn't be too hard to just have two separate ones, one with asan, one without.)

**Comment by silvergasp** [comment]
> I presume that the OSS-Fuzz testing harness sets a limit of time or iterations, since it obviously doesn't run each test forever, so we can just mimic what they do (e.g. --runs=10).

Yeah pretty much, although I think we'll likely want runs to be about `--runs=10000`, when I was testing on my laptop I was getting about ~500 exec/s. So 1000 runs will be a couple seconds of fuzzing. I think replicating what oss-fuzz does would offer limited value. I think the running time is set to about 24 compute hours per day (runs much quicker across a distributed cluster). But without all the corpus management stuff that oss-fuzz has I think there is probably limited value in doing that in CI.

> > Create a cmake toolchain specifically for fuzzing

>That sounds like a fantastic idea.

Easy I'll have a go at that this afternoon, I'm not sure how long this'll take me as I'm far from a cmake expert :)

**Comment by silvergasp** [comment]
Ok I've hopefully addressed most things here in #7566 where I've added;
- Some documentation on fuzz testing, including how to reproduce bugs/crashes found with the fuzzer.
- A fuzz-testing specific toolchain+preset to make building things easier.

Things that I haven't covered/got working so far are;
- Project specific instructions of reproducing oss-fuzz test-cases (I'm going to wait for https://github.com/google/oss-fuzz/pull/10107 to merge first)
- Macos fuzzing builds, as I don't have a mac :(
- Getting fuzz tests running in CI (I'm not really a full bottle on Jenkins, but I could relatively easily do this using a github action).

@steven-johnson you mentioned that you have all of this running internally at google (I assume using clusterfuzz), should I close out https://github.com/google/oss-fuzz/pull/10107 or shall I ping the oss-fuzz team again.

**Comment by steven-johnson** [comment]
> should I close out https://github.com/google/oss-fuzz/pull/10107 or shall I ping the oss-fuzz team again

I think that having it running on OSS-Fuzz (in addition to internally at Google) would be quite valuable, so please continue to ping them :-)
